File: src/org/cerion/symcalc/expression/function/integer/GCD.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class GCD extends FunctionExpr {

	public GCD(Expr ...e) {
		super(FunctionType.GCD, e);
	}
	
	@Override
	protected Expr evaluate() {
		
		//TODO can take more than 2 integer parameters
	
		if(size() == 2 && get(0).isInteger() && get(1).isInteger())
		{
			IntegerNum a = (IntegerNum)get(0);
			IntegerNum b = (IntegerNum)get(1);
			
			return a.gcd(b);
		}
	
		return this;
	}
}

File: src/org/cerion/symcalc/expression/function/integer/GCD.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.exception.ValidationException
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.FunctionExpr
import org.cerion.symcalc.expression.NumberExpr
import org.cerion.symcalc.expression.number.IntegerNum

class GCD(vararg e: Expr) : FunctionExpr(FunctionExpr.FunctionType.GCD, *e) {

    override fun evaluate(): Expr {
        //TODO can take more than 2 integer parameters
        val a = get(0) as IntegerNum
        val b = get(1) as IntegerNum

        return a.gcd(b)
    }

    @Throws(ValidationException::class)
    override fun validate() {
        validateParameterCount(0)
        validateNumberType(0, NumberExpr.INTEGER)
        validateNumberType(1, NumberExpr.INTEGER)
    }
}
